File: src/signer/mock-signer.ts
import type {
  SignerClient,
  PingResult,
  StatusResult,
  GetPublicKeyResult,
  GetPublicKeyCoordinatesResult,
  CreateKeyResult,
  SignHashParams,
  SignHashResult,
  ApproveActionParams,
  ApproveActionResult,
} from './types.js'

// Deterministic test key — never used on-chain
const MOCK_PUBLIC_KEY = '0x04' + 'ab'.repeat(32) + 'cd'.repeat(32)
const MOCK_ADDRESS = ('0x' + 'ee'.repeat(20)) as `0x${string}`
const MOCK_SIGNATURE = ('0x' + 'ff'.repeat(64)) as `0x${string}`
const MOCK_X = ('0x' + 'ab'.repeat(32)) as `0x${string}`
const MOCK_Y = ('0x' + 'cd'.repeat(32)) as `0x${string}`

export function createMockSigner(): SignerClient {
  return {
    async connect() {},
    disconnect() {},

    async ping(): Promise<PingResult> {
      return { pong: true, version: '0.1.0-mock' }
    },

    async status(): Promise<StatusResult> {
      return { ready: true, signerType: 'mock', hasKey: true, publicKey: MOCK_PUBLIC_KEY }
    },

    async getPublicKey(): Promise<GetPublicKeyResult> {
      return { publicKey: MOCK_PUBLIC_KEY, address: MOCK_ADDRESS }
    },

    async getPublicKeyCoordinates(): Promise<GetPublicKeyCoordinatesResult> {
      return { x: MOCK_X, y: MOCK_Y }
    },

    async createKey(): Promise<CreateKeyResult> {
      return { publicKey: MOCK_PUBLIC_KEY, x: MOCK_X, y: MOCK_Y, created: true }
    },

    async signHash(_params: SignHashParams): Promise<SignHashResult> {
      return { signature: MOCK_SIGNATURE, approved: true }
    },

    async approveAction(_params: ApproveActionParams): Promise<ApproveActionResult> {
      return { approved: true }
    },
  }
}

File: signer-daemon/Sources/SignerDaemon/MockSigner.swift
import Foundation

/// Mock signer backend for Stage 0 development.
/// Auto-approves everything. Returns deterministic test keys.
struct MockSigner {
    // Deterministic test key — never used on-chain
    static let publicKey = "0x04" + String(repeating: "ab", count: 32) + String(repeating: "cd", count: 32)
    static let address = "0x" + String(repeating: "ee", count: 20)
    static let signature = "0x" + String(repeating: "ff", count: 64)

    func status() -> StatusResult {
        StatusResult(
            ready: true,
            signerType: "mock",
            hasKey: true,
            publicKey: MockSigner.publicKey
        )
    }

    func getPublicKey() -> GetPublicKeyResult {
        GetPublicKeyResult(
            publicKey: MockSigner.publicKey,
            address: MockSigner.address
        )
    }

    func signHash(_ params: SignHashParams) -> SignHashResult {
        SignHashResult(
            signature: MockSigner.signature,
            approved: true
        )
    }

    func approveAction(_ params: ApproveActionParams) -> ApproveActionResult {
        ApproveActionResult(approved: true, reason: nil)
    }
}
